java/src/com/google/idea/blaze/java/sync/importer/BlazeJavaWorkspaceImporter.java

```diff
@@ -242,24 +242,22 @@ private void addTargetAsSource(
     }
 
     // Add all deps if this target is in the current working set
-    if (workingSet == null || workingSet.isTargetInWorkingSet(target)) {
```

Comment:
This isn't a desirable change for our currently-supported workflows. We don't want to unconditionally add transitive dependencies to the project libraries.

If dependencies which are directly referenced by your project targets aren't being added as libraries during sync, that's a bug specific to your use case -- we'd need to see detailed steps to reproduce to understand why that's happening and how to fix it.

Reply:
I will put together an example.  However, I would submit that the working set should have no impact at all on determining the referenced libraries that should be put into IntelliJ's external libraries list.

Reply:
We could achieve that by removing this block entirely (rather than removing the 'if' statement).

However we have this special-case 'working set' check to support in-progress modifications. If you're currently modifying files in a target, there's a legitimate use-case for unconditionally adding all direct deps without jdeps trimming.

For 'setlled' targets not currently being modified, we rely on the jdeps compilation output to tell us which deps are actually required to resolve the source code, and ignore the rest.

Reply:
I believe the jdeps list was empty - My example was using Kotlin and the rules_kotlin environment.

I will try out both java and Kotlin in the next couple of days to see if there's a difference there.

Reply:
Here is an example repo:  https://github.com/szinn/Bazel-Kotlin

If you import it from Bazel in IntelliJ, the junit jar file won't be there if you have no changed / added files to the repo.  Add a space to test/BUILD and then resync the workspace and the junit external library shows up.  Remove the space which returns you to no altered files, resync, and junit disappears.

Reply:
Thanks for the example. If there are no jdeps produced (or it's an empty file) that would explain what you're seeing.

We rely on jdeps to trim the deps, and assume that it's successfully produced. If the external kotlin rules aren't creating a jdeps file, we'd need to relax that assumption for external kotlin projects.

However this is a serious regression in the external kotlin rules, and something that really should be fixed on the rules side. Without jdeps-based library trimming, IDE performance will scale badly with the size of the kotlin project.

Reply:
I'm taking a closer look now -- doesn't seem to be a problem with the jdeps. I'll update here when it's tracked down.

Reply:
I mistakenly thought the "reading 0 jdeps" meant there weren't any.  However, they are there they just hadn't changed.

In unpacking the jdeps files, it appears that the junit et al are marked as UNUSED rather than EXPLICIT, so that could be a possible issue?  (although I think it unlikely)

Reply:
That is the problem -- junit is marked as unused, so we ignore it (see the 'relevantDep' method in JdepsFileReader). That's definitely a bug in the external kotlin rules.

Reply:
I will dig into that - just tracking down where it actually gets written.  I'll update here when I've found more.  Thanks @brendandouglas !

Reply:
It's not just junit, src/main/greet_lib.jar is also marked as unused, as are all deps other than the kotlin-stdlib ones.

Reply:
It appears to be in the rules-kotlin package.

in JdepsParser.kt, there is code to parse the output of jdep execution to create the .jdeps file.

            val filename = Paths.get(classJar).fileName.toString()
            val jdepsParser = JdepsParser(filename, isImplicit)
            classPath.forEach { x -> jdepsParser.consumeJarLine(x, Deps.Dependency.Kind.UNUSED) }
            lines.forEach { jdepsParser.processLine(it) }

If it isn't one of the Kotlin standard libraries, it gets marked as UNUSED.  

Obviously, that would cause this issue.